Mail redesign

tonijukica opened the pull request:
- Add welcome template
- Add reset template

vladimyr commented on server/common/mail/templates/index.js:
```diff
@@ -0,0 +1,18 @@
+
+function readTemplate(templatePath, params) {
+  const template = fs.readFileSync(path.join(__dirname, templatePath), 'utf8');
+  const icon = { icon: base64Img.base64Sync(path.join(__dirname, iconPath)) };
```
In order to create data uri you need to get file mimetype and base64 encoded contents. To resolve mimetype use [`file-type`](https://npm.im/file-type) that is already added to project deps and for getting base64 encoded contents use `Buffer#toString` with `base64` param 💡 
To combine those two together use: [`create-data-uri`](https://npm.im/create-data-uri)

```js
const createDataUri = require('create-data-uri');
const fileType = require('file-type');

const buf = fs.readFileSync(...);
const { mime } = fileType(buf) || {};
const dataUri = createDataUri(mime, buf.toString('base64'));
```

vladimyr commented on server/common/mail/templates/index.js:
```diff
@@ -0,0 +1,18 @@
+const pupa = require('pupa');
+const base64Img = require('base64-img');
+const mjml2html = require('mjml');
+const iconPath = '/assets/shopping.png';
```
Parse `.mjml` file as html using some _off-the-shelf_ html parser and extract path to image from markup itself instead of hardcoding it here :bowtie:

vladimyr commented on server/common/mail/templates/index.js:
```diff
@@ -0,0 +1,18 @@
+const mjml2html = require('mjml');
+const iconPath = '/assets/shopping.png';
+
+function readTemplate(templatePath, params) {
```
Acutally this is not just _read_ anymore so rename it accordingly :bowtie:

vladimyr commented on server/common/mail/templates/index.js:
```diff
@@ -0,0 +1,18 @@
+const iconPath = '/assets/shopping.png';
+
+function readTemplate(templatePath, params) {
+  const template = fs.readFileSync(path.join(__dirname, templatePath), 'utf8');
```
Don't assume paths relative to `__dirname` here, this function **must** receive fully qualified path from outside.

vladimyr commented on server/common/mail/index.js:
```diff
@@ -25,10 +26,8 @@ function invite(user, { origin }) {
-    Please click <a href="${href}">here</a> to complete your registration.`;
-
+  const recipientName = user.firstName;
+  const message = readTemplate('/assets/welcome.mjml', { href, hostname, recipientName });
```
Use `path.join(__dirname, ...)` here instead.

vladimyr commented on server/common/mail/index.js:
```diff
@@ -41,10 +40,8 @@ function invite(user, { origin }) {
-    Please click <a href="${href}">here</a> to complete the reset process.`;
-
+  const recipientName = user.firstName;
+  const message = readTemplate('/assets/reset.mjml', { href, recipientName });
```
⬆️ Same as above.

vladimyr commented on package.json:
```diff
@@ -31,7 +31,7 @@
   "license": "MIT",
   "devDependencies": {
-    "@babel/core": "^7.1.2",
+    "@babel/core": "^7.2.2",
```
Hm, you shouldn't update stuff you don't really need to because it might introduce bugs elsewhere ⚠️

vladimyr reviewed (changes requested):
Keep up good work 👍

vladimyr commented on server/common/mail/templates/index.js:
```diff
@@ -1,18 +1,25 @@
-  const icon = { icon: base64Img.base64Sync(path.join(__dirname, iconPath)) };
+function getHTML(templatePath, params) {
+  let template = fs.readFileSync(templatePath, 'utf8');
+  const $ = cheerio.load(template, { xmlMode: true });
```
Why `xmlMode`?

vladimyr commented on server/common/mail/templates/index.js:
```diff
@@ -1,18 +1,25 @@
+  template = $.html();
+  const iconBuf = fs.readFileSync(path.join(__dirname, iconPath));
+  const { mime } = fileType(iconBuf) || {};
+  const icon = { icon: createDataUri(mime, iconBuf.toString('base64')) };
```
Extract this to function `getDataUri(imagePath)` :bowtie:

vladimyr commented on package.json:
```diff
@@ -88,6 +89,7 @@
     "fuzzysearch": "^1.0.3",
     "get-stream": "^4.1.0",
     "helmet": "^3.14.0",
+    "html-parse-stringify2": "^2.0.1",
```
You don't use this?

vladimyr commented on server/common/mail/templates/index.js:
```diff
@@ -1,18 +1,25 @@
-function readTemplate(templatePath, params) {
-  const template = fs.readFileSync(path.join(__dirname, templatePath), 'utf8');
-  const icon = { icon: base64Img.base64Sync(path.join(__dirname, iconPath)) };
+function getHTML(templatePath, params) {
```
How about `renderTemplate(templatePath, params)` :bowtie:

tonijukica commented:
Reset mail
![image](https://user-images.githubusercontent.com/37536666/53010442-372fb180-343e-11e9-8b91-621bed0bb5b1.png)
Welcome mail
![image](https://user-images.githubusercontent.com/37536666/53010539-7bbb4d00-343e-11e9-9075-a6ebd6409913.png)

vladimyr commented on package.json:
```diff
@@ -74,7 +74,9 @@
     "bulma": "^0.7.1",
     "bunyan": "^1.8.12",
     "change-case": "^3.0.2",
+    "cheerio": "^1.0.0-rc.2",
```
Umm, please use stable version. ☝️

vladimyr commented on server/common/mail/index.js:
```diff
@@ -23,12 +25,14 @@ module.exports = {
-  const message = `
-    An account has been created for you on ${hostname}.
-    Please click <a href="${href}">here</a> to complete your registration.`;
+  const { protocol, hostname } = new URL(href);
```
How about extracting `origin` and `hostname` instead because `origin` equals `protocol` plus `hostname` 💡

vladimyr commented on server/common/mail/templates/assets/welcome.mjml:
```diff
@@ -0,0 +1,41 @@
+      <mj-column>
+        <mj-text color="#C3C3C3">Welcome <span class="name">{recipientName}</span>,</mj-text>
+        <mj-text color="#C3C3C3">
+          An account has been created for you on <a href="{protocol}//{hostname}">{hostname}</a>.<br />
```
You could use `origin` here instead, look at my previous comment ⬆️

vladimyr commented on server/common/mail/templates/index.js:
```diff
@@ -0,0 +1,30 @@
+const mjml2html = require('mjml');
+const cheerio = require('cheerio');
+const createDataUri = require('create-data-uri');
+const fileType = require('file-type');
```
Sort those imports alphabetically please :bowtie:

vladimyr commented on server/common/mail/templates/index.js:
```diff
@@ -0,0 +1,30 @@
+  $('.icon').attr('src', '{icon}');
+  template = $.html();
+  params = { ...params, icon };
+  const output = pupa(template, params);
```
```suggestion
  const output = pupa($.html(), { ...params, icon });
```

vladimyr commented on server/common/mail/templates/index.js:
```diff
@@ -0,0 +1,30 @@
+  const buf = fs.readFileSync(path.join(__dirname, imagePath));
+  const { mime } = fileType(buf) || {};
+  const icon = createDataUri(mime, buf.toString('base64'));
+  return icon;
```
What's wrong with simply returning result of:
```js
return createDataUri(mime, buf.toString('base64'));
```

vladimyr commented:
Err, is this real font color? Looks too bright? 😕

tonijukica commented:
Welcome mail with new font color
![image](https://user-images.githubusercontent.com/37536666/53022794-31e15f80-345c-11e9-8033-3b2f49f18ece.png)

vladimyr commented:
@tonijukica Please reset your local branch to remote state and squash all those commits to single one using:
```sh
$ git reset -i develop
```
to prevent git history pollution :bowtie:

tonijukica committed:
Add mail templates

vladimyr commented on package.json:
```diff
@@ -20,7 +20,8 @@
     "clean": "del-cli dist/* !dist/.gitkeep",
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "postshrinkwrap": "sed -i.bak 's/http:\\/\\//https:\\/\\//g' package-lock.json && rm -v package-lock.json.bak"
```
This shouldn't be here because we rebased this onto develop that already contains this change.

vladimyr commented on package.json:
```diff
@@ -33,7 +34,7 @@
   "license": "MIT",
   "devDependencies": {
-    "@babel/core": "^7.1.2",
+    "@babel/core": "^7.3.3",
```
Don't change `babel` because you don't need to.

vladimyr commented on package.json:
```diff
@@ -101,9 +104,13 @@
     "jsonwebtoken": "^8.3.0",
+    "jss": "^9.8.7",
+    "jss-camel-case": "^6.1.0",
+    "jss-global": "^3.0.0",
```
We don't use those anymore.

vladimyr commented on server/common/mail/templates/index.js:
```diff
@@ -0,0 +1,50 @@
+const path = require('path');
+const pupa = require('pupa');
+
+const logger = createLogger('mailer:renderer', { level: Level.DEBUG });
```
Actually let's use `mailer` logger but change its level to debug.

vladimyr committed:
Tweak mail template style :art:

- displaying both cta button and actual link
- enlarging cta button
- decreasing footer text size

vladimyr commented:
@kronicker Please review this one and take a look on those new templates:
- [invitation email ✉️ ](https://mailtrap.io/api/v1/inboxes/362261/messages/1160919957/body.html?api_token=6b70ac407f5a3cb12bf653dfac961b37)
- [reset password email ✉️ ](https://mailtrap.io/api/v1/inboxes/362261/messages/1160922158/body.html?api_token=6b70ac407f5a3cb12bf653dfac961b37)

vladimyr committed:
Add plaintext email templates :wrench:

- introducing plaintext mail renderer
- moving mjml templates to templates dir

vladimyr committed:
Cleanup :lipstick:

franosincic commented:
@vladimyr @tonijukica 
Using base64 images in html is maybe not the best idea, based on testing this PR and research, support is still pretty bad in most email clients.
Please read: https://stackoverflow.com/questions/16242489/send-a-base64-image-in-html-email/18646173

zcuric commented:
There is an issues with dynamic data when using `mj-include`. 
For example: If you want to use components with dynamic datat for example `./components/header.mjml` and `./components/footer.mjml`, `pupa` ignores them, as it pulls only the main template and updates the placeholders, i.e. 
```html
<mj-section mj-class="title">
  <mj-column>
    <mj-text mj-class="title" css-class="header" >
      <a href="{origin}">
        <img class="icon" src="{origin}/logo.png" />
      </a>
    </mj-text>
  </mj-column>
</mj-section>
```
and this is inluded in `reset.mjml` template
```
<mj-include path="path/to/templates/components/header.mjml" />
```
In the end `html you'll get this:
```html
<a ... href="{origin}" style="color: #2f72c0; text-decoration: none;">
<img class="icon" src="{origin}/logo.png">
</a>
```
MJML will import as usual.

vladimyr commented:
Sure, sure, didn't think of case when you have multiple source files but that's why we have our heroes @zcuric :+1:

Commit:
Merge branch 'develop' into feature/mail-redesign

Commit:
Removed logo from mail templates

Commit:
Removed icon class

Commit:
Header & footer moved to compontents

Commit:
💄
